clang/lib/Sema/SemaSYCL.cpp

```diff
@@ -3430,147 +3498,134 @@ void Sema::ConstructOpenCLKernel(FunctionDecl *KernelCallerFunc,
   }
 }
 
-void Sema::MarkDevice(void) {
-  // Create the call graph so we can detect recursion and check the validity
-  // of new operator overrides. Add the kernel function itself in case
-  // it is recursive.
-  MarkDeviceFunction Marker(*this);
-  Marker.SYCLCG.addToCallGraph(getASTContext().getTranslationUnitDecl());
-
-  // Iterate through SYCL_EXTERNAL functions and add them to the device decls.
-  for (const auto &entry : *Marker.SYCLCG.getRoot()) {
-    if (auto *FD = dyn_cast<FunctionDecl>(entry.Callee->getDecl())) {
-      if (FD->hasAttr<SYCLDeviceAttr>() && !FD->hasAttr<SYCLKernelAttr>() &&
-          FD->hasBody())
-        addSyclDeviceDecl(FD);
+static void PropagateAndDiagnoseDeviceAttr(Sema &S, Attr *A,
+                                           FunctionDecl *SYCLKernel,
+                                           FunctionDecl *KernelBody) {
+  switch (A->getKind()) {
```

Comment:
I pulled the attribute propagation into its own function here, MarkDevice/MarkDevices was just plain unwieldy as a result.  Most of this is a copy/paste, except it is no longer a member function and has a different tab depth.